internal/grpc/services/ocmshareprovider/ocmshareprovider.go

```diff
@@ -157,6 +159,54 @@ func getOCMEndpoint(originProvider *ocmprovider.ProviderInfo) (string, error) {
 	return "", errors.New("ocm endpoint not specified for mesh provider")
 }
 
+// federatedID creates a federated user id by
+// 1. stripping the protocol from the domain and
+// 2. base64 encoding the opaque id with the domain to get a unique identifier that cannot collide with other users
+func federatedID(id *userpb.UserId) *userpb.UserId {
+	// strip protocol from the domain
+	domain := id.Idp
+	if u, err := url.Parse(domain); err == nil && u.Host != "" {
+		domain = u.Host
+	}
+	return &userpb.UserId{
+		Type:     userpb.UserType_USER_TYPE_FEDERATED,
+		Idp:      domain,
+		OpaqueId: base64.URLEncoding.EncodeToString([]byte(id.OpaqueId + "@" + domain)),
+	}
+}
+
+// remoteID creates a remote user id by
+// 1. decoding the base64 encoded opaque id
+// 2. splitting the opaque id at the last @ to get the opaque id and the domain
+func remoteID(id *userpb.UserId) *userpb.UserId {
+	bytes, err := base64.URLEncoding.DecodeString(id.GetOpaqueId())
+	if err != nil {
```

Comment:
i would like to have

```go
uid := &userpb.UserId{
  Type:     userpb.UserType_USER_TYPE_PRIMARY,
  Idp:      id.Idp,
  OpaqueId: id.OpaqueId,
}
```

in the beginning and than only mutate the OpaqueId if there is no `base64.URLEncoding` err, i think its less code and more easy to understand... personal opinion, up to you